### ISSUE TITLE
history file could be structured data

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#323]
These columns seem almost essential:

- timestamp
- pwd

Also useful:

- elapsed time
- exit status (note: only commands that parsed correctly are even stored?  Or we could store everything)
- evaluated argv (instead of command string, i.e. variables expanded)

https://github.com/oilshell/oil/wiki/Structured-Data-in-Oil

Related to #320 

https://news.ycombinator.com/item?id=20057323

**Comment by ilyash-b** [comment]
More to store in history: https://github.com/ngs-lang/ngs/wiki/History-Design

**Comment by andychu** [comment]
Yeah one problem is that GNU readline does some work for us and it uses it own history format (which is simple, but not easily augmented AFAIK).

I don't want to break Ctrl-R, for example.  I guess the solution is to have two history files.

Does NGS use GNU readline?

BTW my proposal for the textual table format is here:

https://github.com/oilshell/oil/wiki/TSV2-Proposal

This means you can grep the file with plain `grep`, but you can also precisely pick out cells.  You shouldn't have to worry about spaces, tabs, newlines, etc.

**Comment by ilyash-b** [comment]
> Does NGS use GNU readline?

No, there is not REPL yet. I'm also afraid that my vision of UI will not match what readline provides.

> textual table format is here

Quick look reveals that cells must be scalars. That doesn't fit the use case for tables that I was thinking about: AWS API responses.

**Comment by andychu** [comment]
FWIW bash maintainer says that it's difficult in bash because of the GNU readline interface (AFAIU)

https://lists.gnu.org/archive/html/bug-bash/2019-11/msg00019.html
